motionplan/plan.go

```diff
@@ -0,0 +1,154 @@
+package motionplan
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	pb "go.viam.com/api/service/motion/v1"
+	"go.viam.com/rdk/motionplan/ik"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+type Plan struct {
+	Trajectory
+	Path
+
+	// nodes corresponding to inputs can be cached with the Plan for easy conversion back into a form usable by RRT
+	// depending on how the Trajectory is constructed these may be nil and should be computed before usage
+	nodes []node
+}
+
+func newPlan(solution []node, sf *solverFrame, relative bool) (*Plan, error) {
```

Comment:
@nicksanford why is it of benefit to return the object rather than a pointer?

Reply:
A few reasons to prefer the objects themselves rather than pointers to objects in go:

1. Clearer API Design: A pointer type has a less clear semantic meaning than a type itself. Is the pointer expressing optionality i.e. is `nil` a valid parameter or not? Is it the intent for the function to mutate the parameter passed by pointer? All of that is possible & left up to documentation to clarify (which means that on average, it is not very clear). 
2. Safety / Robustness: Whenever you pass pointers to objects in go you need to check the pointer is not nil before using it unless the methods on the pointer expect the pointer to be nil. The consequence of not doing so, is software that will panic when unexpected conditions happen which reduces the robustness & reliability of the software. We very rarely follow that practice of defining methods to operate on nil pointers so in my opinion checking the pointer before calling methods on it is the best way to handle this. B/c pointers need to be checked, which means that any function/method that takes pointer types needs to return an error for the case when one of its pointer parameters is nil. This requires a lot of inconvenient error handling code, which, on average, is then not written(b/c it is inconvenient both to write & to program against) which means that where pointer types are used, there are generally a bunch of places in the code which will panic if nil pointers are passed to them. That makes the software less robust, especially over the lifetime of a software project where today's `nil` pointer is never passed to a function but tomorrow it may be.
3. Performance: Using pointers guarantees the data will need to live on the heap. That is bad for performance for 2 reasons (especially for small objects like this is). Whether passing data by value or by pointer is faster depends on many factors but frequently passing many small objects by pointer in go is bad for performance for the following reasons. Passing pointers reduces the need for copying but frequently a copy (especially of small data structures) can be cheaper than the cost of Cache Misses & GC especially the smaller the objects are & the more of them there are. As a result, I recommend we prefer to pass data by value by default & only pass pointers when we know they are needed for performance:
	1. Cache Misses: It means that it is very unlikely that collections of heap-allocated objects (like slices of pointers to structs) will exist contiguously in memory. Because of that, the CPU will likely need to go to RAM, rather than L1/L2/L3 cache when iterating through a collection of heap-allocated objects, which reduces cache coherency & decreases performance. Stack-allocated objects exist contiguously in memory which is much more likely to fit in cache. 
	2. Garbage Collection Pressure: Heap-allocated objects require more work from the garbage collector. This should be especially of interest to us given that it appears that the largest single contributor of CPU-bound work (20%) done during our most comprehensive automated test of 2d motion planning is GC: https://viam.atlassian.net/browse/RSDK-6309. Stack-allocated objects can be cleaned up when the function they were allocated in terminates. That requires no tracking by the garbage collector & is very cheap. However, heap-allocated objects must be tracked by every time GC runs to see if they have become garbage & need to be deallocated. Passing the data itself means that a copy of the data can live on the stack, which doesn't require any more work from GC.

Reply:
This is a good discussion, but I want to table it for the purposes of this PR.  IMO it should come in another push where we profile the whole stack.